### PR TITLE
Check for expected error codes when running generated JS code

### PR DESCRIPTION
**Pull request by sbc100**
Previously we just disabled the check for returning 0.  Now we
actually check that the program fails.

### CODE REVIEW

File: embuilder.py [hunk 1]
```diff
@@ -58,7 +58,6 @@
 ]
 
 USER_TASKS = [
-    'binaryen',
```
Comment (kripken) [inline embuilder.py:61]: is this an unrelated change?

File: tests/jsrun.py [hunk 2]
```diff
@@ -115,7 +115,11 @@ def run_js(filename, engine=None, args=[],
   out = ['' if o is None else o for o in (stdout, stderr)]
   ret = '\n'.join(out) if full_output else out[0]
 
-  if assert_returncode is not None and proc.returncode is not assert_returncode:
+  # assert_returncode = None means we expect a failure.
+  if assert_returncode is None:
+    if proc.returncode == 0:
```
Comment (kripken) [inline tests/jsrun.py:120]: this surprises me. I'd expect `assert_returncode == None` to mean "don't check it". It seems to instead mean "assert it is not 0"?

Comment (sbc100) [reply tests/jsrun.py:120]: Yeah I agree its confusing... but.  Previously it meant: 

  "we don't care it fails.. i.e. don't check"

Now it means:

 "I expect this to fail, but I don't care about the exact return code"

I experimented with more intrusive changes here but could not find something I liked.    The best alternative I could think of was to have an additional argument call `expect_file=True` which would override the assert_returncode thing.. but this would need to be added in a lot of places.     Perhaps we could have a magic value called ANY_FAIL=xxxx and pass that around?

File: embuilder.py (hunk 1 above)
Comment (sbc100) [reply embuilder.py:61]: Yeah those are part of a separate change now.

Comment (sbc100) [reply embuilder.py:61]: https://github.com/emscripten-core/emscripten/pull/11585

File: tests/jsrun.py (hunk 2 above)
Comment (kripken) [reply tests/jsrun.py:120]: Yeah, a special "fail with any code" value might work, `NON_ZERO` maybe could be the name? then `if assert_returncode is NON_ZERO` would make sense.

Comment (sbc100) [reply tests/jsrun.py:124]: Good idea.  Done.

**Comment by sbc100** [comment]
This should be good to go now..

File: tests/test_core.py [hunk 3]
```diff
@@ -1219,7 +1219,11 @@ def test_exceptions_minimal_runtime(self):
       self.do_run_from_file(path_from_root('tests', 'core', 'test_exceptions.cpp'), path_from_root('tests', 'core', 'test_exceptions_caught.out'))
 
       self.set_setting('DISABLE_EXCEPTION_CATCHING', 1)
-      self.do_run_from_file(path_from_root('tests', 'core', 'test_exceptions.cpp'), path_from_root('tests', 'core', 'test_exceptions_uncaught.out'), assert_returncode=None)
+      # TODO: add assert_returncode=NON_ZERO when node switches its default
```
Comment (kripken) [inline tests/test_core.py:1222]: it's default what?

File: tests/test_core.py [hunk 4]
```diff
@@ -6506,6 +6566,8 @@ def test_zzz_cases(self):
       shortname = os.path.splitext(name)[0]
       # TODO: test only worked in non-fastcomp (well, these cases)
       basename = os.path.basename(shortname)
+      if basename == returns_non_zero[-1]:
```
Comment (kripken) [inline tests/test_core.py:6569]: what does this line do? seems like it's checking `basename == "zeroextarg"` (last element in the array), but that seems weird?

File: tests/test_core.py (hunk 3 above)
Comment (sbc100) [reply tests/test_core.py:1222]: node will currently return 0 for unhandled promise rejection, but the plan on changing that apparanetly.

File: tests/test_core.py (hunk 4 above)
Comment (sbc100) [reply tests/test_core.py:6569]: Oops, didn't mean to commit that.. reverting.

File: tests/test_core.py (hunk 3 above)
Comment (kripken) [reply tests/test_core.py]: Sounds good, a comment maybe?

File: tests/test_benchmark.py [hunk 5]
```diff
@@ -341,7 +341,7 @@ def build(self, parent, filename, args, shared_args, emcc_args, native_args, nat
         try_delete(dir_)
 
   def run(self, args):
-    return jsrun.run_js(self.filename, engine=self.engine, args=args, stderr=PIPE, full_output=True, assert_returncode=None)
+    return jsrun.run_js(self.filename, engine=self.engine, args=args, stderr=PIPE, full_output=True, assert_returncode=jsrun.NON_ZERO)
```
Comment (kripken) [inline tests/test_benchmark.py:344]: we don't test the benchmarks on CI, are you sure this is right? I'm surprised this expects a failure.

File: tests/jsrun.py [hunk 6]
```diff
@@ -115,7 +119,11 @@ def run_js(filename, engine=None, args=[],
   out = ['' if o is None else o for o in (stdout, stderr)]
   ret = '\n'.join(out) if full_output else out[0]
 
-  if assert_returncode is not None and proc.returncode is not assert_returncode:
+  # assert_returncode = None means we expect a failure.
```
Comment (kripken) [inline tests/jsrun.py:122]: this comment is now wrong

File: tests/test_other.py [hunk 7]
```diff
@@ -861,31 +861,31 @@ def test_dangerous_func_cast(self):
...
       run_process([EMCC, '-s', 'WASM=0', 'src.c', '-s', 'ASSERTIONS=1'] + args + ['-s', 'EMULATED_FUNCTION_POINTERS=1'], stderr=PIPE)
-      out = run_js('a.out.js', stderr=PIPE, full_output=True, assert_returncode=None)
+      out = run_js('a.out.js', stderr=PIPE, full_output=True, assert_returncode=assert_returncode)
       self.assertContained(expected, out)
 
     # fastcomp. all asm, so it can't just work with wrong sigs. but,
     # ASSERTIONS=2 gives much better info to debug
     # Case 1: No useful info, but does mention ASSERTIONS
-    test(['-O1'], 'ASSERTIONS')
+    test(['-O1'], 'ASSERTIONS', True)
```
Comment (kripken) [inline tests/test_other.py:882]: what does `assert_returncode=True` mean? is it the same as `1`?

File: tests/test_core.py (hunk 3 above)
Comment (sbc100) [reply tests/test_core.py]: Updated comment

File: tests/test_benchmark.py (hunk 5 above)
Comment (sbc100) [reply tests/test_benchmark.py]: I'm changed to expect success.   All the other benchmark runners seems to expect success so I don't see why chreep wouldn't.   I'm basically removing the ability to be ambivalent about this, so hopefully its easy enough to make sure the cheerp returns zero from these tests.
